aiographql/client/subscription.py

```diff
@@ -42,89 +46,105 @@ def type(self) -> Optional[GraphQLSubscriptionEventType]:
-        return self.message.id
-
-    @property
-    def type(self):
-        return self.message.type
-
-    @property
-    def payload(self):
-        return self.message.payload
-
-
-@dataclass
-class GraphQLSubscription:
-    request: GraphQLRequest
+@dataclass(frozen=True)
+class GraphQLSubscription(GraphQLRequestContainer):
+    id: str = field(default_factory=lambda: str(uuid.uuid4()), init=False)
     callbacks: CallbackRegistry = field(default_factory=CallbackRegistry)
-    id: str = field(default_factory=lambda: str(uuid.uuid4()))
-    headers: Dict[str, str] = field(default_factory=dict)
     stop_event_types: List[GraphQLSubscriptionEventType] = field(
         default_factory=lambda: [
             GraphQLSubscriptionEventType.ERROR,
             GraphQLSubscriptionEventType.CONNECTION_ERROR,
             GraphQLSubscriptionEventType.COMPLETE,
         ]
     )
-    task: Task = None
+    task: Task = field(default=None, init=False, compare=False)
 
     @property
-    def is_running(self):
+    def is_running(self) -> bool:
         return (
             self.task is not None and not self.task.done() and not self.task.cancelled()
         )
 
     @property
-    def is_complete(self):
+    def is_complete(self) -> bool:
         return self.task is not None and (self.task.done() or self.task.cancelled())
 
-    @property
-    def connection_init_request(self):
+    def connection_init_request(self) -> Dict[str, Any]:
         return {
             "type": GraphQLSubscriptionEventType.CONNECTION_INIT.value,
-            "payload": {"headers": self.headers or self.request.headers},
+            "payload": {"headers": {**self.request.headers}},
         }
 
-    @property
-    def connection_start_request(self):
+    def connection_start_request(self) -> Dict[str, Any]:
         return {
             "id": self.id,
             "type": GraphQLSubscriptionEventType.START.value,
-            "payload": self.request.asdict(),
+            "payload": self.request.payload(),
         }
 
-    @property
-    def connection_stop_request(self):
+    def connection_stop_request(self) -> Dict[str, Any]:
         return {"id": self.id, "type": GraphQLSubscriptionEventType.STOP.value}
 
-    def is_stop_event(self, event: GraphQLSubscriptionEvent):
+    def is_stop_event(self, event: GraphQLSubscriptionEvent) -> bool:
         return event.type in self.stop_event_types
 
-    def create_event(self, message: Dict[str, Any]) -> GraphQLSubscriptionEvent:
-        return GraphQLSubscriptionEvent(
-            subscription=self,
-            message=GraphQLSubscriptionMessage(request=self.request, json=message),
-        )
-
     async def handle(self, event: GraphQLSubscriptionEvent) -> NoReturn:
         if event.id is None or event.id == self.id:
             await self.callbacks.handle_event(event.type, event)
+
+    async def _create_websocket_session(self, endpoint: str) -> None:
+        async with aiohttp.ClientSession() as session:
```

Comment:
I did consider this, but it is a bit tricky here I feel. We could move this to being a module level instance I guess? That said, since this is a library I am not sure we should make that a hard decision. Alternatively, we could allow for the session to be passed in maybe? And the session could live within `GraphQLClient`?

Reply:
I like the idea of being able to optionally pass in a session. But even without that, maybe a `get_session` function that initially creates the session and caches it.

Reply:
I have been finding recently keeping the session intact is tricky. But I might do a proposal for this soonish. We can iterate on this later too I guess.